promql/functions.go

```diff
@@ -79,22 +75,29 @@ func funcDelta(ev *evaluator, args Expressions) model.Value {
 		}
 		resultValue := lastValue - samples.Values[0].Value + counterCorrection
 
-		targetInterval := args[0].(*MatrixSelector).Range
-		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp)
-		if sampledInterval == 0 {
-			// Only found one sample. Cannot compute a rate from this.
-			continue
+		// Duration between first/last samples and boundary of range.
+		durationToStart := samples.Values[0].Timestamp.Sub(rangeStart).Seconds()
+		durationToEnd := rangeEnd.Sub(samples.Values[len(samples.Values)-1].Timestamp).Seconds()
+
+		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp).Seconds()
+		averageDurationBetweenSamples := sampledInterval / float64(len(samples.Values))
+
+		// If the first/last samples are close to the boundaries of the range,
+		// extrapolate the result. This is as we expect that another sample
+		// will exist given the spacing between samples we've seen thus far,
+		// with an allowance for noise.
+		extrapolationThreshold := averageDurationBetweenSamples * 1.1
+		extrpolateToInterval := sampledInterval
+		if durationToStart < extrapolationThreshold {
+			extrpolateToInterval += durationToStart
+		}
+		if durationToEnd < extrapolationThreshold {
+			extrpolateToInterval += durationToEnd
+		}
```

Comment:
Our goals are clearly aligned, we just come to opposite conclusions.

I find the jump to extrapolation-free behavior an assumption too bold, and very difficult for the user to figure out, requiring to reverse engineer a lot.

I think with my suggestion we are doing what makes most sense given that Prometheus is a sampling-based system. The case where a data point is for sure the first/last point of a series is pretty special, and my suggestion was for quite some time to provide explicitly extrapolation free functions for that special use-case.


To be clear: This PR solves the worst of problems. Everything we are discussing now is pretty incremental and wouldn't be noticed by most users. I just think we should do things now as correct as possible, based on our current knowledge, and require the smallest set of assumptions, which IMHO requires the median and the half-sampling interval extrapolation for the assumed start and end of a series.